tests/models/test_query_results.py

```diff
@@ -0,0 +1,72 @@
+from redash import models
+from redash.utils import utcnow
+
+
+class QueryResultTest(BaseTestCase):
+    def setUp(self):
+        super(QueryResultTest, self).setUp()
+
+    def test_get_latest_returns_none_if_not_found(self):
+        found_query_result = models.QueryResult.get_latest(self.factory.data_source, "SELECT 1", 60)
+        self.assertIsNone(found_query_result)
+
+    def test_get_latest_returns_when_found(self):
+        qr = self.factory.create_query_result()
+        found_query_result = models.QueryResult.get_latest(qr.data_source, qr.query_text, 60)
+
+        self.assertEqual(qr, found_query_result)
+
+    def test_get_latest_doesnt_return_query_from_different_data_source(self):
+        qr = self.factory.create_query_result()
+        data_source = self.factory.create_data_source()
+        found_query_result = models.QueryResult.get_latest(data_source, qr.query_text, 60)
+
+        self.assertIsNone(found_query_result)
+
+    def test_get_latest_doesnt_return_if_ttl_expired(self):
+        yesterday = utcnow() - datetime.timedelta(days=1)
+        qr = self.factory.create_query_result(retrieved_at=yesterday)
+
+        found_query_result = models.QueryResult.get_latest(qr.data_source, qr.query_text, max_age=60)
+
+        self.assertIsNone(found_query_result)
+
+    def test_get_latest_returns_if_ttl_not_expired(self):
+        yesterday = utcnow() - datetime.timedelta(seconds=30)
+        qr = self.factory.create_query_result(retrieved_at=yesterday)
+
+        found_query_result = models.QueryResult.get_latest(qr.data_source, qr.query_text, max_age=120)
+
+        self.assertEqual(found_query_result, qr)
+
+    def test_get_latest_returns_the_most_recent_result(self):
+        yesterday = utcnow() - datetime.timedelta(seconds=30)
+        old_qr = self.factory.create_query_result(retrieved_at=yesterday)
+        qr = self.factory.create_query_result()
+
+        found_query_result = models.QueryResult.get_latest(qr.data_source, qr.query_text, 60)
+
+        self.assertEqual(found_query_result.id, qr.id)
+
+    def test_get_latest_returns_the_last_cached_result_for_negative_ttl(self):
+        yesterday = utcnow() + datetime.timedelta(days=-100)
+        very_old = self.factory.create_query_result(retrieved_at=yesterday)
+
+        yesterday = utcnow() + datetime.timedelta(days=-1)
+        qr = self.factory.create_query_result(retrieved_at=yesterday)
+        found_query_result = models.QueryResult.get_latest(qr.data_source, qr.query_text, -1)
+
+        self.assertEqual(found_query_result.id, qr.id)
+
+    def test_store_result_does_not_modify_query_update_at(self):
+        original_updated_at = utcnow() - datetime.timedelta(hours=1)
+        query = self.factory.create_query(updated_at=original_updated_at)
+
+        models.QueryResult.store_result(query.org_id, query.data_source, query.query_hash, query.query_text, "", 0, utcnow())
+
+        assert original_updated_at.strftime("%X") == query.updated_at.strftime("%X")
```

Comment:
Please use `self.assertEqual` here and compare `original_updated_at` with `query.updated_at` directly, without formatting it as a string.

Reply:
Isn't `assert` the way you assert when using py.test?

Reply:
Yeah, but as long as we use a unittest base class it's better form to consistently use its assert methods to have a common assertion behavior and error reporting instead of mixing both. pytest transparently supports the unittest assert methods.

Reply:
Wouldn't it be more future proof to use `assert` in new code, even in unittest inheriting classes? This way when we're ready, we can change the base class and don't have to rewrite this test.

Reply:
I'd say that even for the sake of consistency within this module, it makes sense to stick to a single assert. Otherwise it just makes you wonder - what's different about _this_ test?

When we refactor, we'll probably do it as part of a massive search & replace, so one assert doesn't feel like it'll make things easier in the future.

Reply:
> Wouldn't it be more future proof to use `assert` in new code, even in unittest inheriting classes? This way when we're ready, we can change the base class and don't have to rewrite this test.

The compatibility code in pytest for unittest classes is an added layer of abstraction that makes debugging harder when being used side by side with test errors that didn't have the shim applied. In other words, the consistency of using one assertion method (in a module or class) is more important than the decision which to use. When (if?) we move away from unittest classes we can do as @rauchy says, do the conversion in one swap.

Reply:
OK, OK. 🏳️ :-)

I just find the camelCase assert function so ugly that I was happy to stop using them. But of course your arguments make sense (more than my dislike of the camelCase asserts).

Reply:
Oh I believe you and I can assure you the feeling about camelCase asserts is mutual. 😁